File: src/cpp/proxy/proxy.cc
#include <iostream>
#include "proxy.h"

RealSubject::RealSubject()
{
	std::cout << "Constructing a RealSubject" << std::endl;
}

void RealSubject::Request()
{
	std::cout << "Request by RealSubject" << std::endl;
}

Proxy::Proxy()
: m_pRealSubject(NULL)
{
	std::cout << "Constructing a Proxy" << std::endl;
}

Proxy::~Proxy()
{
	delete m_pRealSubject;
	m_pRealSubject = NULL;
}

void Proxy::Request()
{
	if (NULL == m_pRealSubject) {
		std::cout << "Request by Proxy" << std::endl;
		m_pRealSubject = new RealSubject();
	}
	m_pRealSubject->Request();
}

File: src/proxy/php/proxy.php
#!/usr/bin/php
<?php
/**
 * Proxy pattern
 *
 * author:	0xnz
 */

interface Subject {
	public function request();
}

class RealSubject implements Subject {
	public function request() {
		printf("Handling request in real subject\n");
	}
}

class Proxy implements Subject {
	var $subject;

	public function Proxy($subject) {
		$this->subject = $subject;
	}

	// pass request to real subject
	public function  request() {
		$this->subject->request();
	}
}

$proxy = new Proxy(new RealSubject());
$proxy->request();

?>
